### PR TITLE
[FLINK-2283] [streaming] Make grouped reduce/fold/aggregations stateful

### PR DESCRIPTION
**Pull request by mbalassi**
There is an open discussion at the related ticket [1] about fully removing the operators that I touch and partially remove here. I can accept both conclusions of the discussion, but even in the scenario when the operators get removed from the API afterwards the PR has certain merit to it:

1. Cleans up the unused `StreamReduce` and `StreamFold` operators which should be removed either way.
2. Adds an integration test for ensuring that not only user defined functions, but internal streaming operators can properly rely on the `OperatorState` interface. To do this it currently relies on the grouped reduce/fold aggregations, but this is just as important for windowing states, which are not properly checkpointed yet.
3. Makes the grouped fold/reduce operators stateful, so that the previous test can be written.

Some justification for the implementation choices:

1. @gyfora has suggested to use the partitioned state at the ticket [1] instead of the manual map creation. In this scenario the grouped operators would not be unit testable any more as they would be dependent on the state partitioner information found in the keyed datastream. I decided against it.

2. @StephanEwen has recently advised against adding unnecessary integration tests. [2] This is a feature that can only be tested as an integration test. I personally feel the need to cover internal operators with a checkpointing test despite the fact they currently use exactly the same mechanism as the UDFs as this implementation might be subject to slight changes.

3. Elaborating on the previous point the `OperatorState` currently storing the internal state is also accessible for the user. This is an undesirable feature and might lead to accidental overwrite of the state. I am opening a Jira ticket for this. 

[1] https://issues.apache.org/jira/browse/FLINK-2283
[2] https://mail-archives.apache.org/mod_mbox/flink-dev/201509.mbox/%3CCANC1h_vvekciNVDzqCb8N4E5Kfzu4e1Mosnse1%3DV11HXnD2PBQ%40mail.gmail.com%3E

### CODE REVIEW

**Comment by gyfora** [comment]
I still think that we should use PartitionedOperatorStates here, that will also help in case we start thinking about dynamic repartitioning. It's not hard to modify the tests so that they use some partitioning info (so if it unit testable with local state it is unit testable with partitioned state).

**Comment by StephanEwen** [comment]
Is this addressing windowed grouped operations, or the "infinite state" aggregations?

**Comment by aljoscha** [comment]
It is for the "infinite state" aggregations.

I think this will fail if the types (input/output and key type) are not serializable. For those to work a TypeSerializer must be used to but a serialized value into the state map.

**Comment by mbalassi** [comment]
@gyfora: I do not necessarily see that fit together yet, but let us have a brainstorm on the topic.
@StephanEwen: This is the "infinite" version, but I would like to continue with adding it to the windowed versions.
@aljoscha: Thanks for spotting it, that is a problem.

**Comment by gyfora** [comment]
@aljoscha 
A custom StateCheckpointer should be used in this case, there were some issues that I fixed in https://github.com/apache/flink/pull/1154 regarding this.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UdfStreamOperatorCheckpointingITCase.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+				// testing UDF folder
+				.fold(Tuple2.of(0, 0L), new FoldFunction<Tuple2<Integer, Long>, Tuple2<Integer, Long>>() {
+					@Override
+					public Tuple2<Integer, Long> fold(
+							Tuple2<Integer, Long> accumulator, Tuple2<Integer, Long> value) throws Exception {
+						return Tuple2.of(value.f0, accumulator.f1 + value.f1);
+					}
+				})
+				.groupBy(0)
+				.addSink(new FoldEvictingQueueSink());
```
Comment (tillrohrmann) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UdfStreamOperatorCheckpointingITCase.java]: Formatting looks off on Github

**Comment by mbalassi** [comment]
Quick update:
I made some progress on the custom state checkpointer solution, but run into an issue with key selectors. This would break the current API usage of aggregations, so I need to solve it or work around it first. [1]

[1] https://issues.apache.org/jira/browse/FLINK-2812

**Comment by StephanEwen** [comment]
@marton: I am making an attempt to make the state backend abstraction a bit more powerful, as described here: https://issues.apache.org/jira/browse/FLINK-2808

**Comment by mbalassi** [comment]
@StephanEwen : Thanks for the heads-up.

Now I have addressed the comments and would like to merge this tomorrow if there are no objections.

**Comment by StephanEwen** [comment]
I think it is fine. As I see it, your changes only use the key/value state, which will continue to be surfaced in the same way, so it should work together fine.

**Comment by StephanEwen** [comment]
@mbalassi This is a valid case for an integration test!

**Comment by mbalassi** [comment]
@StephanEwen: Thanks, I will go ahead and merge it.
I still had a small hick-up in the scala api, that I had fixed just now - so I am waiting for travis.

**Pull request merged**